### ISSUE TITLE
Infinite recursion error in one to many with inversion

### ISSUE THREAD

**Issue by tflori** [laravel/framework#29758]
- Laravel Version: 5.x (We are using 5.4 but it should also fail in 5.8)
- PHP Version: 7.x
- Database Driver & Version: irrelevant

### Description:
`$model->push()` can create infinite recursions when the related entity has a relation to itself.

### Steps To Reproduce:
Assume you have an Article with many comments. The comment also defines a "belongsTo" Article. 

```php
// first we load an article with at least one comment
$article = Article::query()->has('comments')->first();
// now we load the comments and remember the first one
$comment = $article->comments->first();
// we load the article of the comment - for what ever reason; could be a trigger somewhere
$comment->article;

// we duplicate this comment and push it to save also the article
$duplicate = $comment->replicate();
$duplicate->push();
```

In https://github.com/laravel/framework/blob/5.8/src/Illuminate/Database/Eloquent/Model.php#L610 is defined the push method will save the comment and then iterate over all relations and push them too. One of the relations is the article and the article has the comment in its relations and the comment has the article in its relations and the article has the comment in its relation and .. you see that is a bad idea.

### Solution

When we pass all parents that are already pushed we can check that we don't run into a recursion.

```php
public function push($parents = []) {
  $this->save();
  foreach ($this->relations as $models) {
    foreach ((array)$models as $model) {
      if (in_array($model, $parents)) {
        continue;
      }
      $model->push(array_merge($parents, [$this]));
    }
  }
}
```

**Comment by driesvints** [comment]
Please verify this on 5.8 first. Thanks.

**Comment by tflori** [comment]
I don't see a reason to verify it on laravel 5.8. I can't update my application because of too much breaking changes. You will have to verify it on your own. The code has not changed since them.

There is a method that calls itself on another object without any checks for recursion. For sure this fails. If you don't want to fix it, fine! All models in our application have a parent class in our namespace - and laravel 5.4 will not be fixed anyway. So if I want to fix it in our application I have to do it anyway on our behalf.

It's the 3rd or 4th issue that just got closed without a solution or anything else. :confused:

**Comment by driesvints** [comment]
@tflori we don't support 5.4 anymore so please verify this on 5.8 first.

**Comment by rimace** [comment]
I'll be happy to have a look and work on this, but probably not in time for 5.8 but rather something like 6.0.x.
@tflori If you need that fix and currently can't update to a higher version I may have a workaround for you. You could clone this repo, use it instead and then back-port the (to be provided) fix to version 5.4.
Hope that helps for now.
